upstairs/src/client.rs

```diff
@@ -0,0 +1,3133 @@
+        std::mem::forget(client_stop_rx);
+
+        self.client_task = Some(ClientTaskHandle {
+            client_request_tx,
+            client_stop_tx: Some(client_stop_tx),
+            client_response_rx,
+        });
+    }
+    /// Starts a client IO task, saving the handle in `self.client_task`
+    ///
+    /// If we are running unit tests and `self.target_addr` is not populated, we
+    /// start a dummy task instead.
+    ///
+    /// # Panics
+    /// If `self.client_task` is not `None`, or `self.target_addr` is `None` and
+    /// this isn't running in test mode
+    fn start_task(&mut self, delay: bool) {
+        #[cfg(test)]
+        if self.target_addr.is_none() {
+            self.start_dummy_task();
+        } else {
+            self.start_network_task(delay);
+        }
+
+        #[cfg(not(test))]
+        self.start_network_task(delay);
+    }
+
+    fn start_network_task(&mut self, delay: bool) {
+        assert!(
+            self.client_task.is_none(),
+            "cannot start task when it is already running"
+        );
+        let target = self.target_addr.expect("socket address is not populated");
+
+        // These channels must support at least MAX_ACTIVE_COUNT messages;
+        // otherwise, we risk a deadlock if the IO task and main task
+        // simultaneously try sending each other data when the channels are
+        // full.
+        let (client_request_tx, client_request_rx) =
+            mpsc::channel(MAX_ACTIVE_COUNT * 2);
+        let (client_response_tx, client_response_rx) =
+            mpsc::channel(MAX_ACTIVE_COUNT * 2);
+        let (client_stop_tx, client_stop_rx) = oneshot::channel();
+
+        let tls_context = self.tls_context.clone();
+        let log = self.log.new(o!("" => "io task"));
+        let client_id = self.client_id;
+        tokio::spawn(async move {
+            client_run(
+                client_id,
+                tls_context,
+                target,
+                client_request_rx,
+                client_stop_rx,
+                client_response_tx,
+                delay,
+                log,
+            )
+            .await
+        });
+        self.reset_timeout();
+        self.client_task = Some(ClientTaskHandle {
+            client_request_tx,
+            client_stop_tx: Some(client_stop_tx),
+            client_response_rx,
+        });
+    }
+
+    /// Indicate that the upstairs has requested that we go active
+    ///
+    /// This either sent a `PromoteToActive` request directly, or schedules it
+    /// to be sent once the client state reaches `WaitActive`.
+    ///
+    /// # Panics
+    /// If we already called this function (without `reinitialize` in between),
+    /// or `self.state` is invalid for promotion.
+    pub(crate) async fn set_active_request(&mut self, gen: u64) {
+        match self.promote_state {
+            Some(PromoteState::Waiting(..)) => {
+                panic!("called set_active_request while already waiting")
```

Comment:
Yes; this was intentional to match [the behavior of existing code](https://github.com/oxidecomputer/crucible/blob/4fe7be7b4c9ce668ab13677659caa79cf8484171/upstairs/src/lib.rs#L5787-L5788).

We could change it, though!

Reply:
Let's change it to avoid some future debug time.

Reply:
Okay, done in 2e7865af0982d0e1a14003e30c67d43ca1978c95 (reusing the existing `CrucibleError::UpstairsAlreadyActive`, to avoid changing the public error type)